Soroban metrics improvement

jayz22 opened the pull request:
# Description

Resolves all the items in this checklist: https://github.com/stellar/stellar-core-internal/issues/236#issuecomment-1915049419

Most of them are just some naming and types change. There is one structural change:

A new `SorobanLedgerMetrics` was added in `LedgerManager` which tracks soroban resource usage per ledger across all soroban ops (invoke_host_fn, restore_footprint, extend_ttl), and publishes them to new metrics "soroban.ledger.xxx".

<!---

Describe what this pull request does, which issue it's resolving (usually applicable for code changes).

--->

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

MonsieurNicolas commented on docs/metrics.md:
```diff
@@ -166,39 +166,49 @@ soroban.host-fn-op.read-entry                | meter     | number of entries rea
 soroban.host-fn-op.read-key-byte             | meter     | number of key bytes in read entries
 soroban.host-fn-op.write-key-byte            | meter     | number of key bytes in written entries
-soroban.host-fn-op.read-ledger-byte          | meter     | number of data + code bytes in read entries
+soroban.host-fn-op.read-ledger-byte          | meter     | number of entry (data + code) bytes in read entries
```
wording is weird (it was confusing before too), should it just be number of bytes read?

jayz22 replied:
Fixed.

MonsieurNicolas commented on docs/metrics.md:
```diff
@@ -166,39 +166,49 @@ soroban.host-fn-op.read-entry                | meter     | number of entries rea
 soroban.host-fn-op.read-data-byte            | meter     | number of data bytes in read entries
 soroban.host-fn-op.read-code-byte            | meter     | number of code bytes in read entries
-soroban.host-fn-op.write-ledger-byte         | meter     | number of data + code bytes in written entries
+soroban.host-fn-op.write-ledger-byte         | meter     | number of entry (data + code) bytes in written entries
```
same wrt writes -- maybe just "number of bytes written for Soroban entries"

jayz22 replied:
Fixed.

MonsieurNicolas commented on docs/metrics.md:
```diff
@@ -166,39 +166,49 @@ soroban.host-fn-op.read-entry                | meter     | number of entries rea
-soroban.host-fn-op.cpu-insn-excl-vm          | meter     | metered cpu instructions excluding VM instantation
-soroban.host-fn-op.invoke-time-nsecs-excl-vm | meter     | time [nsecs] spent in `invoke_host_function` excluding VM instantation
+soroban.host-fn-op.invoke-time-nsecs         | timer     | time spent in `invoke_host_function`
+soroban.host-fn-op.cpu-insn-excl-vm          | meter     | metered cpu instructions excluding VM instantiation
```
should specify -- is this per `invoke_host_function` op?

MonsieurNicolas replied:
as in, if a single `invoke_host_function`  op calls multiple VMs, is it recorded as one or multiple events?

MonsieurNicolas replied:
(we probably want 1 event per op)

jayz22 replied:
I've cleaned up this section. Yes it is per-op.

MonsieurNicolas commented on docs/metrics.md:
```diff
@@ -166,39 +166,49 @@ soroban.host-fn-op.read-entry                | meter     | number of entries rea
-soroban.host-fn-op.success                   | meter     | if `InvokeHostFunctionOp` results in a success
-soroban.host-fn-op.failure                   | meter     | if `InvokeHostFunctionOp` results in a failure
+soroban.host-fn-op.success                   | meter     | number of successful `InvokeHostFunctionOp` operations
+soroban.host-fn-op.count                     | meter     | number of total `InvokeHostFunctionOp` operations
```
we don't need this metric (total count): count is already recorded in any other metric that tracks `InvokeHostFunctionOp` data, like `soroban.host-fn-op.invoke-time-nsecs`

jayz22 replied:
Ah yes, we didn't have it before (everything before was "meter"). Now we indeed don't need the count. I've reverted that commit (and keeps the "failure" metric).

MonsieurNicolas commented on docs/metrics.md:
```diff
@@ -166,39 +166,49 @@ soroban.host-fn-op.read-entry                | meter     | number of entries rea
-soroban.config.contract-data-key-size-bytes      | meter     | soroban config setting `contract_data_key_size_bytes`
-soroban.config.contract-data-entry-size-bytes    | meter     | soroban config setting `contract_data_entry_size_bytes`
\ No newline at end of file
+soroban.restore-fprint-op.read-ledger-byte   | meter     | number of entry bytes in read entries
```
fix wording, it's hard to understand (same for others in this group)

jayz22 replied:
Fixed.

MonsieurNicolas commented on docs/metrics.md:
```diff
@@ -166,39 +166,49 @@ soroban.host-fn-op.read-entry                | meter     | number of entries rea
+soroban.ledger.read-byte                     | histogram | number of entry bytes in read entries per ledger
+soroban.ledger.write-entry                   | histogram | number of entries written per ledger
+soroban.ledger.write-byte                    | histogram | number of entry bytes in written entries per ledger
+soroban.config.contract-max-rw-key-byte      | counter   | soroban config setting `contract_data_key_size_bytes`
```
what is the point of exporting those soroban.config.* as metrics?

MonsieurNicolas replied:
(I agree they should be `counter` if we keep them)

jayz22 replied:
In order to display the limits on top of all of the resource consumption metrics (defined above). Without these we'll have to hardcode them on grafana and display them as variables, which is a static and inconvenient.

jayz22 commented:
@MonsieurNicolas thanks for the review, and sorry about the confusion. Most of the confusion comes from "read" == both read and modified, and "write" == only modified, and "entry" == `LedgerEntry` which could be {ContractDataEntry, ContractCodeEntry}, and we also have "data" and "code" which are just those two variants.  
I've cleaned up the wording of the entire section, hope that makes it clear now.

graydon commented:
r+ fed81d0
